rpc/accounts.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/0xsequence/waas-authenticator/rpc/attestation"
+	"github.com/0xsequence/waas-authenticator/rpc/crypto"
+	"github.com/0xsequence/waas-authenticator/rpc/tenant"
+)
+
+func (s *RPC) listAccounts(
+	ctx context.Context, sess *data.Session, intent *intents.IntentTyped[intents.IntentDataListAccounts],
+) ([]*intents.Account, error) {
+	tntData := tenant.FromContext(ctx)
+
+	dbAccounts, err := s.Accounts.ListByUserID(ctx, sess.UserID)
+	if err != nil {
+		return nil, fmt.Errorf("listing DB accounts: %w", err)
+	}
+
+	out := make([]*intents.Account, len(dbAccounts))
+	for i, dbAcc := range dbAccounts {
+		accData, _, err := crypto.DecryptData[*proto.AccountData](ctx, dbAcc.EncryptedKey, dbAcc.Ciphertext, tntData.KMSKeys)
+		if err != nil {
+			return nil, fmt.Errorf("decrypting account data: %w", err)
+		}
+
+		var identity proto.Identity
+		if err := identity.FromString(accData.Identity); err != nil {
+			return nil, fmt.Errorf("parsing account identity: %w", err)
+		}
+
+		out[i] = &intents.Account{
+			ID:     identity.String(),
+			Type:   intents.IdentityType(identity.Type.String()),
+			Issuer: identity.Issuer,
+		}
+		if dbAcc.Email != "" {
+			out[i].Email = &dbAcc.Email
+		}
+	}
+	return out, nil
+}
+
+func (s *RPC) federateAccount(
+	ctx context.Context, sess *data.Session, intent *intents.IntentTyped[intents.IntentDataFederateAccount],
+) (*intents.Account, error) {
+	att := attestation.FromContext(ctx)
+	tntData := tenant.FromContext(ctx)
+
+	if intent.Data.SessionID != sess.ID {
+		return nil, fmt.Errorf("sessionId mismatch")
+	}
+
+	sessionHash := ethcoder.Keccak256Hash([]byte(strings.ToLower(sess.ID))).String()
+	identity, err := verifyIdentity(ctx, s.HTTPClient, intent.Data.IdToken, sessionHash)
+	if err != nil {
+		return nil, fmt.Errorf("verifying identity: %w", err)
+	}
+
+	_, found, err := s.Accounts.Get(ctx, tntData.ProjectID, identity)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving account: %w", err)
+	}
+	if found {
+		return nil, fmt.Errorf("account already exists")
+	}
+
+	accData := &proto.AccountData{
+		ProjectID: tntData.ProjectID,
+		UserID:    sess.UserID,
+		Identity:  identity.String(),
+		CreatedAt: time.Now(),
+	}
+
+	encryptedKey, algorithm, ciphertext, err := crypto.EncryptData(ctx, att, tntData.KMSKeys[0], accData)
+	if err != nil {
+		return nil, fmt.Errorf("encrypting account data: %w", err)
+	}
+
+	account := &data.Account{
+		ProjectID:          tntData.ProjectID,
+		Identity:           data.Identity(identity),
+		UserID:             accData.UserID,
+		Email:              identity.Email,
+		ProjectScopedEmail: fmt.Sprintf("%d|%s", tntData.ProjectID, identity.Email),
+		EncryptedKey:       encryptedKey,
+		Algorithm:          algorithm,
+		Ciphertext:         ciphertext,
+		CreatedAt:          accData.CreatedAt,
+	}
+
+	if _, err := s.Wallets.FederateAccount(waasContext(ctx), account.UserID, convertToAPIIntent(intent.ToIntent())); err != nil {
```

Comment:
What will happen if this call is success, however we fail to put account in accounts on authenticator side? Will we end up in recoverable state?

Reply:
Waas-Auth assumes this call is idempotent so we will always be in a recoverable state - I ensured this is the case on the API side